File: openclaw_republic/agents/executive/sec_state.py
"""国务卿 Agent (Secretary of State) — 外部信息检索与 API 交互。

挂载 WebBrowser、Search 技能，
负责外部信息检索与 API 交互。
"""

from __future__ import annotations

from typing import Any

from openclaw_republic.agents.base import BaseAgent, Branch, Permission
from openclaw_republic.schemas.act import ExecutionTask, TaskResult


class SecretaryOfState(BaseAgent):
    """国务卿 — 负责外部信息获取。"""

    _available_tools = ["WebBrowser", "Search"]

    def __init__(self) -> None:
        super().__init__(
            name="Sec. of State",
            role="sec_state",
            branch=Branch.EXECUTIVE,
            permissions={Permission.EXECUTE},
        )

    async def execute_task(self, task: ExecutionTask) -> TaskResult:
        """执行总统派发的检索任务。

        当前阶段为 Mock 实现 — 校验权限与工具可用性后
        返回成功的 ``TaskResult``。真实工具调用在 Phase 2+ 实现。

        Args:
            task: 总统分派的执行任务。

        Returns:
            执行结果。

        Raises:
            PermissionDeniedError: 无 EXECUTE 权限。
            ValueError: 所需 Skill 不在本部长可用工具集中。
        """
        self.require_permission(Permission.EXECUTE)

        if not self.can_use_tool(task.step.required_skill):
            raise ValueError(
                f"{self.role} 无法使用工具 '{task.step.required_skill}'，可用工具: {self._tools}"
            )

        # Mock 执行 — 模拟消耗 estimated_tokens
        return TaskResult(
            task_id=task.task_id,
            step_index=task.step.index,
            status="success",
            output=f"[Mock] {self.role} 完成步骤 {task.step.index}: {task.step.description}",
            tokens_consumed=task.step.estimated_tokens,
        )

    # ----- BaseAgent.act() 实现 -----

    async def act(self, message: Any) -> Any:
        """核心处理入口。

        Args:
            message: 预期为 ``ExecutionTask`` 实例。

        Returns:
            ``TaskResult``。
        """
        if isinstance(message, ExecutionTask):
            return await self.execute_task(message)
        raise TypeError(f"{self.role} 不接受 {type(message).__name__} 类型消息")

File: backend/src/agents/executive/sec-state.ts
import { BaseAgent, Branch, Permission } from '../base';
import { EventAction } from '../../schemas/events';
import { ActStep, ExecutionTask, TaskResult } from '../../schemas/act';
import { OpenClawAdapter } from '../../openclaw/adapter';
import { MessageBus } from '../../bus/message-bus';
import { truncateOutput } from '../../openclaw/sandbox';

export class SecretaryOfState extends BaseAgent {
  protected static _available_tools: string[] = ['WebBrowser', 'Search'];

  constructor(adapter: OpenClawAdapter, bus?: MessageBus) {
    super('Sec. of State', 'sec_state', Branch.EXECUTIVE, [Permission.EXECUTE], adapter, bus, true);
  }

  // ── Prompt Construction ─────────────────────────────────────────────────

  /**
   * Build a role-specific prompt for the given step.
   * Search → summarized search results; WebBrowser → page content extraction.
   */
  public _buildTaskPrompt(step: ActStep): string {
    switch (step.required_skill) {
      case 'Search':
        return `你现在是国务卿（Secretary of State）。请使用你的搜索工具（Search）完成以下任务。
返回搜索到的关键信息摘要，确保信息准确且相关。

<task_description>
${step.description}
</task_description>

要求：
- 使用搜索工具查找相关信息
- 返回精炼的结果摘要（不超过 2000 字）
- 如果搜索无果，明确说明`;

      case 'WebBrowser':
        return `你现在是国务卿（Secretary of State）。请使用你的浏览器工具（WebBrowser）完成以下任务。

<task_description>
${step.description}
</task_description>

要求：
- 使用浏览器工具访问相关页面
- 提取并返回页面的关键内容
- 如果页面无法访问，说明原因`;

      default:
        return `请完成以下任务：\n<task_description>\n${step.description}\n</task_description>`;
    }
  }

  // ── Main Execution ──────────────────────────────────────────────────────

  public async executeTask(task: ExecutionTask): Promise<TaskResult> {
    this.requirePermission(Permission.EXECUTE);

    if (!this.canUseTool(task.step.required_skill)) {
      throw new Error(
        `${this.role} 无法使用工具 '${task.step.required_skill}'，可用工具: ${this._tools.join(', ')}`
      );
    }

    this.emitEvent(EventAction.TOOL_CALL, {
      tool_name: task.step.required_skill,
      step_index: task.step.index,
      status: 'running',
    }, undefined, task.task_id);

    try {
      const prompt = this._buildTaskPrompt(task.step);
      const result = await this.callLLM(prompt);

      // Empty content → treat as failure
      if (!result.content || !result.content.trim()) {
        this.emitEvent(EventAction.TOOL_CALL, {
          tool_name: task.step.required_skill,
          step_index: task.step.index,
          status: 'failed',
          error: 'LLM returned empty response',
        }, undefined, task.task_id);

        return {
          task_id: task.task_id,
          step_index: task.step.index,
          status: 'failed',
          output: '',
          tokens_consumed: 0,
          error: 'LLM returned empty response',
        };
      }

      const output = truncateOutput(result.content);

      this.emitEvent(EventAction.TOOL_CALL, {
        tool_name: task.step.required_skill,
        step_index: task.step.index,
        status: 'success',
        output: output,
      }, undefined, task.task_id);

      return {
        task_id: task.task_id,
        step_index: task.step.index,
        status: 'success',
        output: output,
        tokens_consumed: task.step.estimated_tokens,
      };
    } catch (err: unknown) {
      const errorMessage = err instanceof Error ? err.message : String(err);
      // Emit failure event
      this.emitEvent(EventAction.TOOL_CALL, {
        tool_name: task.step.required_skill,
        step_index: task.step.index,
        status: 'failed',
        error: errorMessage,
      }, undefined, task.task_id);

      // Return structured failure — never block the Pipeline
      return {
        task_id: task.task_id,
        step_index: task.step.index,
        status: 'failed',
        output: '',
        tokens_consumed: 0,
        error: errorMessage,
      };
    }
  }

  public async act(message: unknown): Promise<unknown> {
    const msg = message as ExecutionTask;
    if (msg && msg.task_id && msg.step && msg.assigned_to === this.role) {
      return await this.executeTask(msg);
    }
    throw new TypeError(`${this.role} 不接受此类型消息`);
  }
}
